### DISCUSSION TITLE
[ZK Stack] - How to run ZkSync Node for a self hosted ZkSync Hyperchain?

### DISCUSSION THREAD

**Discussion by sachanayush47** [zkSync-Community-Hub/zksync-developers#695] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [ ] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [x] Other (please specify below)

### Rust Version

1.80.1

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

I have deployed ZkSync Hyperchain on my AWS EC2 instance connect to L1 Sepolia. Now Im trying to run a ZkSync Node to it. 
I found this tutorial https://docs.zksync.io/zksync-node/quickstart to setup ZkSync Node locally but it is configured to setup with zksync's mainnet and testnet. How can one run a node for a self hosted hyperchain?

How do I get the consensus config? What all other changes are required

### Current Behavior

Error Im getting after running testnet-external-node-docker-compose.yml.
I have updated the EN_ETH_CLIENT_URL, EN_MAIN_NODE_URL and EN_L2_CHAIN_ID envs.

```
Error: One or more tasks failed: [TaskFailed(TaskId("oneshot_runner"), Oneshot task node_storage_initializer failed
external-node-1     | 
external-node-1     | Caused by:
external-node-1     |     0: snapshot recovery failed
external-node-1     |     1: no snapshots on main node; snapshot recovery is impossible)]
external-node-1     | 2024-09-19T10:28:00.934718Z  INFO zksync_external_node: Main node URL is: "https://subdomain.domain.com:1111/"
external-node-1     | 2024-09-19T10:28:01.360711Z  INFO zksync_external_node::node_builder: Pruning is disabled
external-node-1     | 2024-09-19T10:28:01.361550Z  INFO zksync_utils::env: locate_workspace() failed. You are using an already compiled version
```

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
To run a ZKsync Node for a self-hosted ZKsync Hyperchain on your AWS EC2 instance connected to L1 Sepolia, you'll need to adapt the standard setup instructions for local or testnet environments to your specific configuration. Here’s a general guide on how to proceed:

1. **Clone the ZKsync Era Repository**: Start by cloning the ZKsync Era repository to your server.
   ```bash
   git clone https://github.com/matter-labs/zksync-era.git
   cd zksync-era
   ```

2. **Navigate to the Node Setup Guide**: After cloning the repository, navigate to the node setup guide within the docs.
   ```bash
   cd docs/guides/external-node
   ```

3. **Modify Configuration for Sepolia**: Since the standard setup is for the mainnet or testnet, you will need to modify the Docker Compose files or configuration files to point to your L1 Sepolia setup. This typically involves changing RPC URLs and possibly other network-specific parameters.

4. **Start the Node**: Use Docker Compose to start your node. You might need to create a custom Docker Compose file based on your configuration.
   ```bash
   docker compose up
   ```

5. **Monitor and Adjust**: Monitor the logs and performance of your node. Adjust configurations as needed to ensure stable connectivity and performance with your L1 Sepolia connection.

For detailed instructions and configurations specific to a self-hosted setup, you might need to refer to additional resources or community support, as most documentation focuses on connecting to the predefined networks like ZKsync's mainnet or testnet.

For further guidance, you can also check the [Quick Start Guide for ZKsync Node](https://docs.zksync.io/zksync-node/quickstart) which provides foundational steps that might be adapted for your setup.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by sachanayush47** [reply]
How do I get the consensus config?

**Reply by alhenaai** [reply]
To obtain the consensus configuration for your ZKsync node, especially when setting up a node for a self-hosted ZKsync Hyperchain, you will typically find this information within the main configuration files of your node setup. Here’s how you can access and modify these configurations:

1. **Access Configuration Files**: The consensus configurations are usually part of the main node configuration files. For a ZKsync node, these files are often located in the directory where you have installed or cloned the ZKsync node software.

2. **Locate the Consensus Config Section**: Within the configuration files, look for sections labeled similarly to `consensus`, `chain`, or `network`. These sections contain parameters that govern the node's behavior in terms of block validation, block production, and other consensus-related activities.

3. **Edit as Necessary**: If you need to customize the consensus parameters (like adjusting timeouts, block sizes, or other consensus rules), you can edit these directly within the configuration files. Be sure to follow any specific guidelines provided in the documentation to avoid misconfigurations.

4. **Documentation and Help**: Refer to the official ZKsync documentation or community forums for specific guidance on consensus configurations. The documentation often provides detailed descriptions of each configuration parameter and its impact on the node's operation.

If you are using a standard setup from the ZKsync documentation, such as the ones found at [ZKsync Node Configuration](https://docs.zksync.io/zksync-node/configuration), you might need to adapt the instructions based on your specific requirements for a self-hosted Hyperchain.

For more detailed and specific instructions, it would be beneficial to consult the documentation related to the version of the ZKsync software you are using, as consensus mechanisms can vary significantly between different versions and setups.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by sachanayush47** [reply]
The error im getting

Error: One or more tasks failed: [TaskFailed(TaskId("oneshot_runner"), Oneshot task node_storage_initializer failed
 
Caused by:
     0: snapshot recovery failed
     1: no snapshots on main node; snapshot recovery is impossible)]

**Reply by alhenaai** [reply]
The error you're encountering indicates that the node storage initializer task failed because it couldn't perform snapshot recovery—specifically, it couldn't find any snapshots on the main node to recover from. Here's how you can address this issue:

### Steps to Resolve the Error

1. **Check Snapshot Settings**:
   - Ensure that your node configuration is set to either generate snapshots or that it's correctly pointed to a source that provides snapshots. This setting is crucial if your node setup relies on snapshots for state recovery.

2. **Verify Main Node Configuration**:
   - Confirm that the main node (which your node might be trying to sync from) is configured to create and store snapshots. If the main node is within your control, check its settings to ensure snapshots are being generated and stored as expected.

3. **Manual Snapshot Import**:
   - If snapshots are not available on the main node, you might need to manually import a snapshot from another source or use a database dump if available. This can be done by placing the snapshot files in the appropriate directory and configuring your node to use these files for initialization.

4. **Disable Snapshot Recovery**:
   - As a temporary measure, you can disable snapshot recovery in your node's configuration. This would allow the node to start without relying on external snapshots, though it might increase the initialization time as the node will need to sync from the genesis block or a specified block height.

5. **Consult Documentation**:
   - Review the [ZKsync Node Troubleshooting Guide](https://docs.zksync.io/zksync-node/troubleshooting) for specific instructions on handling snapshot issues. This guide might provide additional steps or configurations that are specific to your version of the ZKsync node.

6. **Logs and Monitoring**:
   - Check the node's logs for any additional errors or warnings that might give more context about why the snapshot recovery is failing. This can help in pinpointing configuration issues or network problems that might be affecting snapshot accessibility.

### Example Configuration Change
If you decide to disable snapshot recovery temporarily, you might need to adjust a configuration line in your node's setup file from something like:
```yaml
enable_snapshots_recovery: true
```
to:
```yaml
enable_snapshots_recovery: false
```

After making any changes, restart your node to apply the new settings. If the problem persists, consider reaching out to the ZKsync community or support channels for more targeted assistance.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by sachanayush47** [reply]
Please raise this with the team. You are not helpful.

**Reply by alhenaai** [reply]
I'm sorry to hear that the information provided wasn't helpful for your specific situation. Since I'm unable to directly raise issues with the team, I recommend reaching out through the official channels for more direct assistance.

You can discuss this issue further with the ZKsync team and community by:
- Posting in the [Github Discussion forum](https://github.com/zkSync-Community-Hub/zksync-developers/discussions/categories/announcements)
- Contacting support directly via [email]

These channels are monitored by the ZKsync development team who can provide more detailed and technical support.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by bxpana** [comment]
Will see if this is possible right now. Thanks for the feedback on the AI responses. Right now there isn't a ton of documentation out there for ZK Stack so the responses for ZK Stack aren't always full. 

BTW are you from a specific team or are you just looking into things?

**Reply by sachanayush47** [reply]
We are trying to build something specific. Earlier we are using polygon, but due to lack of dev support and we have moved to zksync. Dev experience of zksync is also a lot much better than polygon. Easy to understand docs, more stable, etc.

**Comment by RomanBrodetski** [comment]
hey, yes this should be possible and the AI was right to adopt that configuration (you can take [this](https://github.com/matter-labs/zksync-era/blob/main/docs/guides/external-node/docker-compose-examples/testnet-external-node-docker-compose.yml#L96) as example and modify the highlighted and potentially other fields.

You should also disable snapshot recovery [here](https://github.com/matter-labs/zksync-era/blob/main/docs/guides/external-node/docker-compose-examples/testnet-external-node-docker-compose.yml#L103) since - I assume - you don't generate snapshots on your env.

Please let us know if it works!

**Reply by sachanayush47** [reply]
After applying your suggested changes, Im getting this new error:

```
 Error: One or more tasks failed: [TaskFailed(TaskId("oneshot_runner"), Oneshot task l1_batch_commitment_mode_validation failed
external-node-1     | 
external-node-1     | Caused by:
external-node-1     |     Configured L1 batch commitment mode (Rollup) does not match the commitment mode used on L1 contract 0xadd82b0c015be3fb7e8410061b4f15d577b4207f (Validium))]
```

Couple of more questions:
1. How can I generate snapshots?
2. Testnet YML file configured to use [Testnet consensus config](https://github.com/matter-labs/zksync-era/blob/e9d1d905f1ce86f9de2cf39d79be4b5aada4a81d/docs/guides/external-node/docker-compose-examples/testnet-external-node-docker-compose.yml#L108). Does this need any changes?


Edit:
Above contract address belongs to
diamond_proxy_addr: 0xadd82b0c015be3fb7e8410061b4f15d577b4207f

**Reply by bxpana** [reply]
@sachanayush47 checking to see if you still were having issues with this.

**Reply by sachanayush47** [reply, marked as answer]
> @sachanayush47 checking to see if you still were having issues with this.

Yeah, This is fixed. I found out that the correct way to deploy a external node is by using zk_inception/zkstack cli not via example docker files.
